community/kernel/src/test/java/org/neo4j/kernel/api/index/IndexEntryUpdateTest.java

```diff
@@ -117,4 +117,21 @@ public void changedShouldRetainValues() throws Exception
         assertThat( multiValue, equalTo( multiChange.beforeValues() ) );
         assertThat( multiAfter, equalTo( multiChange.values() ) );
     }
+
+    @Test
+    public void indexEntryUpdatesShouldBeEqualForDeepValues()
+    {
+        Object value1 = arraysOfDepth( 13 );
+        Object value2 = arraysOfDepth( 13 );
+
+        IndexEntryUpdate a = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), value1 );
+        IndexEntryUpdate b = IndexEntryUpdate.add( 0, NewIndexDescriptorFactory.forLabel( 3, 4 ), value2 );
+        assertThat( a, equalTo( b ) );
+        assertThat( a.hashCode(), equalTo( b.hashCode() ) );
+    }
+
+    private Object[] arraysOfDepth( int n )
+    {
+        return n == 0 ? new Object[]{"hej"} : arraysOfDepth( n-1 );
```

Comment:
This doesn't wrap or nest any arrays. It just returns the array at the recursion base case.